clippy_lints/src/loops/manual_find.rs

```diff
@@ -0,0 +1,198 @@
+                            &format!(
+                                ".map(|{}| {})",
+                                snippet_with_applicability(cx, pat.span, "..", &mut applicability),
+                                ident
+                            )[..],
+                        );
+                    }
+                    let mut refs = String::new();
+                    for _ in 0..ref_level {
+                        refs.push('&');
+                    }
+                    snippet.push_str(
+                        &format!(
+                            ".find(|{}{}| {})",
+                            refs,
+                            ident,
+                            snippet_with_applicability(cx, cond.span, "..", &mut applicability)
+                        )[..],
+                    );
+                    if ref_level > 1 {
+                        for _ in 0..(ref_level - 1) {
+                            // In `for &el in arr`, arr elements must be `Copy`
+                            snippet.push_str(".copied()");
+                        }
+                    }
+                    span_lint_and_sugg(
+                        cx,
+                        MANUAL_FIND,
+                        span.with_hi(base_case.span.hi()),
+                        "manual implementation of Iterator::find",
+                        "replace with an iterator",
+                        snippet,
+                        applicability,
+                    );
+                }
+            }
+        }
+    }
+}
+
+fn ident_of(expr: &Expr<'_>) -> Option<Ident> {
+    if let ExprKind::Path(QPath::Resolved(_, path)) = expr.kind {
+        if let [seg, ..] = path.segments {
+            return Some(seg.ident);
+        }
+    }
+    None
+}
+
+fn ident_in_expr<'tcx>(cx: &LateContext<'tcx>, ident: Ident, expr: &'tcx Expr<'_>) -> bool {
+    let mut contains = false;
+    expr_visitor(cx, |e| {
+        contains |= ident_of(e).map_or(false, |i| i == ident);
+        !contains
+    })
+    .visit_expr(expr);
+    contains
+}
+
+#[derive(Default)]
+struct PatIdentCollector {
+    idents: Vec<(Ident, usize)>,
+}
+
+impl PatIdentCollector {
+    fn collect(&mut self, pat: &Pat<'_>) {
+        self.visit_pat(pat, 0);
+    }
+
+    fn visit_pat(&mut self, pat: &Pat<'_>, ref_level: usize) {
+        match pat.kind {
+            PatKind::Binding(_, _, ident, _) => self.idents.push((ident, ref_level)),
+            PatKind::Struct(_, fields, _) => {
+                for field in fields {
+                    self.visit_pat(field.pat, ref_level);
+                }
+            },
+            PatKind::TupleStruct(_, pats, _) | PatKind::Or(pats) | PatKind::Tuple(pats, _) => {
+                for pat in pats {
+                    self.visit_pat(pat, ref_level);
+                }
+            },
+            PatKind::Box(p) => self.visit_pat(p, ref_level),
+            PatKind::Ref(p, _) => self.visit_pat(p, ref_level + 1),
```

Comment:
This doesn't quite work when references cross the boundary of another Pat, e.g. ref -> tuple -> ref, examples:

```rust
fn tuple_with_ref(v: [(u8, &u8); 3]) -> Option<u8> {
    for (_, &x) in v {
        if x > 10 {
            return Some(x);
        }
    }
    None
}
```
```rust
fn ref_to_tuple_with_ref(v: &[(u8, &u8)]) -> Option<u8> {
    for &(_, &x) in v {
        if x > 10 {
            return Some(x);
        }
    }
    None
}
```